source/common/router/router.cc

```diff
@@ -777,22 +825,29 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
   // already.
   ASSERT(buffering || !upstream_requests_.empty());
 
+  // We will need to make N copies of the data.
+  int num_copies = buffering +                   // one copy for the buffer.
+                   !upstream_requests_.empty() + // one copy for the main request.
+                   shadow_streams_.size();       // one copy for each shadow.
+  auto many_copied_buffer = Buffer::ManyCopiedBuffer(data, num_copies);
+
+  if (!upstream_requests_.empty()) {
+    upstream_requests_.front()->acceptDataFromRouter(many_copied_buffer.nextBuffer(), end_stream);
+  }
   if (buffering) {
-    // If we are going to buffer for retries or shadowing, we need to make a copy before encoding
-    // since it's all moves from here on.
-    if (!upstream_requests_.empty()) {
-      Buffer::OwnedImpl copy(data);
-      upstream_requests_.front()->acceptDataFromRouter(copy, end_stream);
+    callbacks_->addDecodedData(many_copied_buffer.nextBuffer(), true);
+  }
+  if (streaming_shadows_ && !shadow_streams_.empty()) {
+    for (auto* shadow_stream : shadow_streams_) {
+      if (end_stream) {
+        shadow_stream->removeDestructorCallback();
+        shadow_stream->removeWatermarkCallbacks();
+      }
+      shadow_stream->sendData(many_copied_buffer.nextBuffer(), end_stream);
+    }
+    if (end_stream) {
+      shadow_streams_.clear();
```

Comment:
Why clear the streams at this point? Wont it all get torn down when the filter is destroyed?

Reply:
It's more of a semantic thing -- I want `shadow_streams_` to only contain streams to which have not seen `end_stream=true`.